### PR TITLE
Require KYC to enroll in promo registrations

### LINKED ISSUE #2730: Bare Bones Referral KYC Requirements
* Restrict new users from enrolling in the referral program without KYC
* Beginning of work to phase out referrals

### PR DESCRIPTION
**Pull request by corymcdonald**
## Require KYC to enroll in promo registrations

Closes #2730

#### Features

* adds welcome modal

### CODE REVIEW

File: app/views/admin/publishers/show.html.slim [hunk 1]
```diff
@@ -118,12 +118,14 @@ hr
           hr
 
   .c-4.shadow-sm.bg-white.rounded.p-3.my-4
-    .panel
-      input type="checkbox" name="panels" id="panel_id"
-        label for="panel_id"
-          h3.text-dark Add Note
-      .panel-content
-        = render partial: 'admin/publisher_notes/form', locals: { note: PublisherNote.new, publisher: @publisher }
+    .accordion
```
Comment (corymcdonald) [inline app/views/admin/publishers/show.html.slim:121]: Unrelated change :) We were just missing a CSS class here that I added.

File: app/models/publisher.rb [hunk 2]
```diff
@@ -323,8 +332,37 @@ def paypal_locale?(locale)
...
+  def valid_promo_country?
+    PromoRegistration::RESTRICTED_COUNTRIES.exclude?(country)
+  end
+
+  def promo_registrable?
+    # If the user doesn't have the referral_kyc_flag on then we can register them still.
+    return true unless referral_kyc_required?
+
+    # Otherwise they must be brave payable and from a valid country
+    (brave_payable? && valid_promo_country?) || (promo_expiration_time.present? && promo_not_expired?)
```
Comment (corymcdonald) [inline app/models/publisher.rb:354]: This shoudl only contain one check otherwise promo expiration time won't hold

File: app/services/payout/potential_payment/uphold_service.rb [hunk 3]
```diff
@@ -54,29 +55,31 @@ def perform
...
+        wallet_provider: PotentialPayment.wallet_providers['uphold'],
+        suspended: @publisher.suspended?,
+        status: @publisher.last_status_update&.status,
+        channel_stats: channel.details.stats,
+        channel_type: channel.details_type
+      )
+    end
+
+    unless should_only_notify?
+      potential_payments.each(&:save)
```
Comment (corymcdonald) [inline app/services/payout/potential_payment/uphold_service.rb:82]: Review with eV: Write out a payout message and update some metabase dashboards to display this result.

File: app/models/publisher.rb [hunk 4]
```diff
@@ -323,8 +332,37 @@ def paypal_locale?(locale)
...
+    provider_country = uphold_connection&.country || paypal_connection&.country
+
+    provider_country.to_s.upcase
+  end
+
+  def valid_promo_country?
+    PromoRegistration::RESTRICTED_COUNTRIES.exclude?(country)
+  end
+
+  def promo_registrable?
```
Comment (yachtcaptain23) [inline app/models/publisher.rb:349]: May we name this `may_register_promo?`

File: db/migrate/20200529195058_add_promo_expired_to_publishers.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+class AddPromoExpiredToPublishers < ActiveRecord::Migration[6.0]
+  def change
+    add_column :publishers, :promo_expiration_time, :datetime
```
Comment (yachtcaptain23) [inline db/migrate/20200529195058_add_promo_expired_to_publishers.rb:3]: Wouldn't it be easier if we kept `promo_expiration_time` inside the `feature_flags` column?

File: app/models/publisher.rb (hunk 4 above)
Comment (corymcdonald) [reply app/models/publisher.rb]: 👍 https://github.com/brave-intl/publishers/pull/2733/commits/99a33d783c7fcaa136639550816a0fcc9bd39936

File: app/models/publisher.rb (hunk 2 above)
Comment (corymcdonald) [reply app/models/publisher.rb]: https://github.com/brave-intl/publishers/pull/2733/commits/99a33d783c7fcaa136639550816a0fcc9bd39936

File: app/javascript/locale/en.ts [hunk 6]
```diff
@@ -31,6 +31,20 @@ export default {
...
+        description: "Follow the instructions to verify that you're the rightful owner of properties.",
+        number: "1",
+        title: "Add your websites or channels",
+      },
+      second: {
+        description: "To receive payments, your wallet accounts must be connected. Connect Uphold for your desired payment options.",
+        number: "2",
+        title: "Connect wallet services",
+      },
+      title: "Weclome to Brave Rewards",
```
Comment (evq) [inline app/javascript/locale/en.ts:47]: *welcome

File: app/services/payout/potential_payment/uphold_service.rb [hunk 7]
```diff
@@ -28,9 +29,9 @@ def perform
     total_probi = probi
 
     # Create the referral payment for the owner
-    unless should_only_notify?
-      PotentialPayment.create(
-        payout_report_id: @payout_report.id,
+    if @publisher.may_register_promo?
```
Comment (evq) [inline app/services/payout/potential_payment/uphold_service.rb:32]: I think we're missing the case where you can no longer register but your already active promo codes have not yet expired

File: db/migrate/20200529195058_add_promo_expired_to_publishers.rb (hunk 5 above)
Comment (yachtcaptain23) [reply db/migrate/20200529195058_add_promo_expired_to_publishers.rb:3]: "expiration" might not be the best word, perhaps "lockout" is a better term e.g. "promo_lockout_time"

Comment (corymcdonald) [reply db/migrate/20200529195058_add_promo_expired_to_publishers.rb]: https://github.com/brave-intl/publishers/pull/2733/commits/8aba6249557ebce7ad13a023acde4121cb6d127a

https://github.com/brave-intl/publishers/pull/2733/commits/179360eac96f50570ecbd2957162cfe566b24417

File: app/javascript/locale/en.ts (hunk 6 above)
Comment (corymcdonald) [reply app/javascript/locale/en.ts]: Fixed the translations! Also added Japanese locales https://github.com/brave-intl/publishers/pull/2733/commits/b301a96874ed81adf0e72261f6d10b6df3c644f5

File: app/services/payout/potential_payment/uphold_service.rb (hunk 7 above)
Comment (corymcdonald) [reply app/services/payout/potential_payment/uphold_service.rb]: Agreed 👍  I added it here along with some tests

https://github.com/brave-intl/publishers/pull/2733/commits/c5add56e31a711592192fe6a08cb7463b1b43afa

File: app/services/payout/potential_payment/uphold_service.rb (hunk 3 above)
Comment (corymcdonald) [reply app/services/payout/potential_payment/uphold_service.rb]: https://github.com/brave-intl/publishers/pull/2733/commits/2ed9d5a73a67dba6a923d9521d98e045d2af9a6a

File: app/models/concerns/user_feature_flags.rb [hunk 8]
```diff
@@ -15,17 +25,28 @@ module UserFeatureFlags
...
+      # If the field is a checkbox then we can cast it to a boolean
       # False is "0" and anything else is 0 for HTML forms
-      checked = update_flag_params[flag_param_key] != "0"
-      if checked.present?
-        feature_flags[flag_param_key] = true
+      if value == "0" || value.include?("checked")
+        value = ActiveModel::Type::Boolean.new.cast(value)
+      end
+
+      puts '🐱🐱🐱🐱🐱🐱🐱🐱🐱🐱'
```
Comment (evq) [inline app/models/concerns/user_feature_flags.rb:36]: nyaa

Comment (corymcdonald) [reply app/models/concerns/user_feature_flags.rb]: ope https://github.com/brave-intl/publishers/pull/2733/commits/cd36a7d25c7c12344eb20e03c6660d2f00903dac

File: app/models/publisher.rb [hunk 9]
```diff
@@ -254,6 +254,15 @@ def promo_status(promo_running)
     end
   end
 
+  def promo_expired?
```
[the remark this answers is not in the document]
Comment (corymcdonald) [reply app/models/publisher.rb]: https://github.com/brave-intl/publishers/pull/2733/commits/cd36a7d25c7c12344eb20e03c6660d2f00903dac
